paddle/fluid/extension/include/op_function.h

```diff
@@ -0,0 +1,296 @@
+/* Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License. */
+
+#pragma once
+
+#include <functional>
+#include <iostream>
+#include <string>
+#include <tuple>
+#include <typeindex>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "paddle/fluid/framework/tensor.h"
+#include "paddle/fluid/platform/place.h"
+
+/**
+ * Op Function Related Define.
+ *
+ * Used to maintain operator core information independent of the framework.
+ *
+ */
+
+namespace paddle {
+
+using Tensor = framework::Tensor;
+
+#define DISABLE_COPY_AND_ASSIGN(classname)         \
+ private:                                          \
+  classname(const classname&) = delete;            \
+  classname(classname&&) = delete;                 \
+  classname& operator=(const classname&) = delete; \
+  classname& operator=(classname&&) = delete
+
+//////////////////// Kernel Function traits (PD_TRAITS) ///////////////////
+
+// Record forward function traits
+using FuncInfo = std::pair<size_t, size_t>;
+using TraitsFunc = FuncInfo (*)();
+
+template <typename T, typename Enabled = void>
+struct KernelFuncTraits;
+
+template <typename Return, typename... Args>
+struct KernelFuncTraits<Return (*)(Args...)> {
+  static FuncInfo GetFuncInfo() {
+    // TODO(chenweihang): parse tensor args num & attribute num
+    // Now only Tensor is input by default
+    return std::make_pair(Arity, 0);
+  }
+
+ private:
+  using ReturnType = Return;
+  using ArgsTuple = std::tuple<Args...>;
+
+  enum : std::size_t { Arity = sizeof...(Args) };
+
+  template <std::size_t Index>
+  using Arg = typename std::tuple_element<Index, ArgsTuple>::type;
+};
+
+#define OP_INFO(...) \
+  ::paddle::KernelFuncTraits<decltype(&__VA_ARGS__)>::GetFuncInfo
+
+////////////////////// Kernel Function (PD_KERNEL) ////////////////////////
+
+// Record Op kernel core function
+using KernelFunc = std::vector<Tensor> (*)(std::vector<const Tensor*> inputs,
+                                           std::vector<boost::any> attrs);
+
+template <typename T>
+struct TypeTag {};
+
+template <typename F, F f>
+struct KernelFuncImpl;
+
+template <typename Return, typename... Args, Return (*impl_fn)(Args...)>
+struct KernelFuncImpl<Return (*)(Args...), impl_fn> {
+  static Return Compute(std::vector<const Tensor*> inputs,
+                        std::vector<boost::any> attrs) {
+    return ComputeCallHelper<Args..., TypeTag<int>>::template Compute<0, 0>(
```

Comment:
Using idx size to end the recursive calling of this instead of TypeTag is bettert

Reply:
Here we have two compile time index, we can get total size of two index, but not each idx size